Added solver for 2nd order autonomous ODE

naveensaigit opened the pull request:
Added the solver for 2nd order nonlinear autonomous ODE.

#### References to other Issues or PRs
Fixes #19362 

#### Brief description of what is fixed or changed
Solutions with unevaluated integral fails in `is_special_case_of`. So, I have added a check to make them pass. Let me know if that is not the right thing to do.

#### Other comments


#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
   * Added solver for 2nd order nonlinear autonomous ODE. 
<!-- END RELEASE NOTES -->

naveensaigit committed:
feat(dsolve): Added solver for 2nd order autonomous ODEs

oscarbenjamin commented:
The ODE module is currently being refactored (work done mostly by @Mohitbalwani26). The new design is to have solvers as classes like this:
https://github.com/sympy/sympy/blob/5288eb336d12e4f3c3070bce67d75f38ab4d8376/sympy/solvers/ode/single.py#L378-L449

Any new solvers should implemented in the new style.

naveensaigit commented:
> The ODE module is currently being refactored (work done mostly by @Mohitbalwani26). The new design is to have solvers as classes like this:

Oh, I didn't know that. Could you review the changes? It should not take too long to change it into a class

Mohitbalwani26 commented:
> Oh, I didn't know that. Could you review the changes? It should not take too long to change it into a class. 

I guess it would be better if you could directly start implementing with class based approach as later on anyway it should be in that way.

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -3693,6 +3709,51 @@ def ode_2nd_linear_bessel(eq, func, order, match):
+    g = g.subs(f(x), u)
+    C1, C2 = get_numbered_constants(eq, num=2)
+    inside = -2*Integral(g, (u, f(x))).doit().subs(f(x), u) + C1
+    lhs = Integral(1/sqrt(inside), (u, f(x))).doit()
```
This should not use doit. It's better to return the formula in unevaluated integrals. Later on `dsolve` will call `doit` but the user should have the option to not call `doit` using:
```
hint = '2nd_nonlinear_autonomous_conserved_Integral'
```

naveensaigit replied:
We could do that. I notice that, before returning the solutions, the helper functions simplify and remove redundant solutions. However, equations of this kind usually have an unevaluated integral in the solution and it is taking forever to simplify this, and in some cases fails after 10-15 mins. Can I just skip simplification just for this hint? Or else, the solver would practically be useless.

naveensaigit replied:
Actually skipping simplification would be bad as some solutions are being simplified. Is there any other solution to this?

oscarbenjamin replied:
Perhaps in general if `simplify=False` then we should skip the remove redundant solutions step. Or otherwise perhaps remove redundant solutions should only attempt to remove cases where the solutions are linear in the constants.

Otherwise we can improve the way that solutions are simplified if it is being slow.

naveensaigit replied:
Setting `simplify=False` is not always good as some equations which can be simplified don't get simplified.
I noticed that the simplification is only slow if there is an integral that can't be evaluated. Is there any way to recursively check if an expression contains an `Integral` object? Then, we can check if the hint is `2nd_nonlinear_autonomous_conserved` and if the solution has an Integral, return it as it is.

naveensaigit replied:
> Otherwise we can improve the way that solutions are simplified if it is being slow.

The call stack looks too complicated. I'm not sure if this would work

oscarbenjamin replied:
Yes, but for now the user can try `dsolve(..., simplify=False)` if `dsolve` is slow. We can test some examples with `simplify=False`.

I'm sure that the simplification in `dsolve` can be improved.

naveensaigit replied:
Ok

oscarbenjamin commented:
It would be good to come up with some better examples where this method works if possible (more examples that don't need `dsolve_too_slow` or `checkodesol_XFAIL`).

naveensaigit commented:
> It would be good to come up with some better examples where this method works if possible (more examples that don't need `dsolve_too_slow` or `checkodesol_XFAIL`).

I tried many examples, but the problem here is that solutions are often too complicated due to the equation being nonlinear. I'll see if there's anything I missed. I could add the case where the function is a polynomial in `y` (Constant Coefficients), but I didn't think this would be useful. Let me know if I have to add this.

oscarbenjamin commented:
This doesn't work in the general case:
```
In [2]: dsolve(f(x).diff(x, 2)-g(f(x)), f(x))
---------------------------------------------------------------------------
ValueError 
```

naveensaigit commented:
> This doesn't work in the general case:
> 
> ```
> In [2]: dsolve(f(x).diff(x, 2)-g(f(x)), f(x))
> ---------------------------------------------------------------------------
> ValueError 
> ```

I have fixed this, will add it in the next commit.

Mohitbalwani26 commented on sympy/solvers/ode/tests/test_single.py:
```diff
@@ -809,8 +822,8 @@ def _get_examples_ode_sol_factorable():
-        'sol': [], #currently dsolve doesn't return any solution for this example
-        'XFAIL': ['factorable']
+        'sol': [Eq(f(x), log(C1/(cos(C1*sqrt(-1/C1)*(C2 + x)) - 1))), Eq(f(x), log(C1/(cos(C1*sqrt(-1/C1)*(C2 - x)) - 1))), Eq(f(x), log(C1/(1 - cos(C1*sqrt(-1/C1)*(C2 + x))))), Eq(f(x), log(C1/(1 - cos(C1*sqrt(-1/C1)*(C2 - x)))))],
+        'dsolve_too_slow': True,
```
Is this solution returned by dsolve with hint `Factorable`? If not, then move this example under this newly added solver examples and there keep `XFAIL` as key for Factorable as it was earlier.

naveensaigit replied:
Yes, `Factorable` simplifies the equation and the method I have added gives one of the solutions to it. Previously, `dsolve` was unable to compute the solution, but now it can do it, so I have added the corresponding solution

oscarbenjamin replied:
I think factorable factorises the ODE and then calls dsolve again. The new solver here is then able to return an answer for the factorised ODE which was previously not solvable.

naveensaigit replied:
Yes

oscarbenjamin replied:
Really `factorable` shouldn't be a "solver" hint. It should just be part of the preprocessing of `dsolve`. That's something that we can fix after `dsolve` is refactored.

naveensaigit commented:
@oscarbenjamin The test that fails is this-
``` python
eq = f(x).diff(x, 2) - x*f(x)
classify_ode(eq) #Matches with 2nd_nonlinear_autonomous_conserved
```
I'm not sure if this should match. I think it shouldn't. I tried to find the solution using the same hint, but `checkodesol` hangs.
``` python
>>> eq=f(x).diff(x,2)-x*f(x)
>>> dsolve(eq, hint='2nd_nonlinear_autonomous_conserved', simplify=False)
[Eq(-asinh(sqrt(x)/sqrt(C1))/sqrt(x) + asinh(sqrt(x)*f(x)/sqrt(C1))/sqrt(x), C2 + x),
Eq(-asinh(sqrt(x)/sqrt(C1))/sqrt(x) + asinh(sqrt(x)*f(x)/sqrt(C1))/sqrt(x), C2 - x)]
>>> checkodesol(eq, _)
#Hangs
```
If this is wrong, how can I stop this from being matched? I tried to include `x` in the `exclude` parameter of the `Wild` object during the match, but then its not able to match even terms like `g(f(x))`, `exp(f(x))`, etc

oscarbenjamin commented:
> I'm not sure if this should match. I think it shouldn't.

It shouldn't match as it's not autonomous.

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -708,7 +710,8 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
     if isinstance(rv, list):
-        rv = _remove_redundant_solutions(eq, rv, order, func.args[0])
+        if hint != "2nd_nonlinear_autonomous_conserved" and hint != "2nd_nonlinear_autonomous_conserved_Integral" or simplify:
+            rv = _remove_redundant_solutions(eq, rv, order, func.args[0])
```
We shouldn't check for specific hints here.

I think just checking for simplify should be fine if that means we can test the solver. The real fix is to make it so that `_remove_redundant_solutions` only handles simple cases e.g. those that are linear in the constants.

naveensaigit replied:
Okay, will remove these

oscarbenjamin commented:
You could add a `_verify` method.

naveensaigit commented:
> It shouldn't match as it's not autonomous. You could add a `_verify` method.

I'll do that

naveensaigit commented:
I think this is good to be merged. Please review @oscarbenjamin

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+    >>> eq = f(x).diff(x, 2) + sqrt(f(x))
+    >>> dsolve(eq, simplify=False)
+    [Eq(2*f(x)*gamma(2/3)*hyper((1/2, 2/3), (5/3,), 4*exp_polar(2*I*pi)*f(x)**(3/2)/(3*C1))/(3*sqrt(C1)*gamma(5/3)) - 2*gamma(2/3)*hyper((1/2, 2/3), (5/3,), 4*exp_polar(2*I*pi)/(3*C1))/(3*sqrt(C1)*gamma(5/3)), C2 + x),
+    Eq(2*f(x)*gamma(2/3)*hyper((1/2, 2/3), (5/3,), 4*exp_polar(2*I*pi)*f(x)**(3/2)/(3*C1))/(3*sqrt(C1)*gamma(5/3)) - 2*gamma(2/3)*hyper((1/2, 2/3), (5/3,), 4*exp_polar(2*I*pi)/(3*C1))/(3*sqrt(C1)*gamma(5/3)), C2 - x)]
```
These lines are far too long.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+    >>> eq = f(x).diff(x, 2) - exp(f(x)) + log(f(x))
+    >>> dsolve(eq, simplify=False)
+    [Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u + 2*exp(u)), (u, 1, f(x))), C2 + x),
+    Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u + 2*exp(u)), (u, 1, f(x))), C2 - x)]
```
A better example for the docstring would be the general case:
```python
In [3]: s1, s2 = dsolve(f(x).diff(x, 2) - g(f(x)), simplify=False)

In [4]: s1
Out[4]: 
f(x)                                   
 ⌠                                     
 ⎮             1                       
 ⎮   ────────────────────── du = C₂ + x
 ⎮       __________________            
 ⎮      ╱        ⌠                     
 ⎮     ╱  C₁ + 2⋅⎮ g(u) du             
 ⎮   ╲╱          ⌡                     
 ⌡                                     
 1                                     

In [5]: s2
Out[5]: 
f(x)                                   
 ⌠                                     
 ⎮             1                       
 ⎮   ────────────────────── du = C₂ - x
 ⎮       __________________            
 ⎮      ╱        ⌠                     
 ⎮     ╱  C₁ + 2⋅⎮ g(u) du             
 ⎮   ╲╱          ⌡                     
 ⌡                                     
 1    
```

naveensaigit replied:
Okay, I'll replace the example that's too long with this.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+        C1, C2 = self.ode_problem.get_numbered_constants(num=2)
+        inside = -2*Integral(g, u) + C1
+        lhs = Integral(1/sqrt(inside), (u, 1, fx))
+        return [Eq(lhs, C2 + x), Eq(lhs, C2 - x)]
```
Why does the integral go from 1?

naveensaigit replied:
This is something I saw in Wolfram, the integral is calculated from 1 to `f(x)`. Should I remove it?

naveensaigit replied:
Should I just make it `Integral(1/sqrt(inside), (u, fx))` instead of `Integral(1/sqrt(inside), (u, 1, fx))`?

oscarbenjamin replied:
Yes, I think so.

oscarbenjamin commented on sympy/solvers/ode/tests/test_single.py:
```diff
@@ -377,6 +377,18 @@ def test_2nd_2F1_hypergeometric_intrgral():
+    sol = [Eq(Integral(1/sqrt(C1 - 2*Integral(asin(u), u)), (u, 1, f(x))), C2 + x),
+    Eq(Integral(1/sqrt(C1 - 2*Integral(asin(u), u)), (u, 1, f(x))), C2 - x)]
+    assert sol == dsolve(eq, hint='2nd_nonlinear_autonomous_conserved_Integral', simplify=False)
+    assert checkodesol(eq, [s.doit() for s in sol]) == [(True, 0), (True, 0)]
```
Why is the doit here needed. We should improve `checkodesol` to handle this case.

naveensaigit replied:
I'll look into it.

oscarbenjamin commented on sympy/solvers/ode/tests/test_single.py:
```diff
@@ -1751,6 +1764,35 @@ def _get_examples_ode_sol_2nd_2F1_hypergeometric():
+            '2nd_nonlinear_autonomous_conserved_01': {
+                'eq': f(x).diff(x, 2) + exp(f(x)) + log(f(x)),
+                'sol': [Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u - 2*exp(u)), (u, 1, f(x))), C2 + x), Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u - 2*exp(u)), (u, 1, f(x))), C2 - x)],
+                'simplify_flag': False,
```
Some of the lines here are too long and should be formatted more nicely.

It is better not to fully indent these (like the other tests)

naveensaigit replied:
I'll split it up into 2 lines.

naveensaigit replied:
Do you want me to split the tests into lines like this test?
https://github.com/sympy/sympy/blob/4c20029861a0cdb8ab2356b6f57cc8f8d5b2b8b4/sympy/solvers/ode/tests/test_single.py#L1868-L1874
Is there any reason why the tests shouldn't be indented? It looks cleaner with the indent IMO.

oscarbenjamin replied:
We want to stop the line length being too long. Adding more indentation means that more equations have to be wrapped across lines.

I'm guessing that you are asking about this because your editor is trying to force the indentation or something. We could use standard indentation rules like:
```python
    examples = {}
    example_set = {
        'hint': "2nd_nonlinear_autonomous_conserved",
        'func': f(x),
        'examples': examples,
    }
    examples['2nd_nonlinear_autonomous_conserved_01'] = {
        'eq': f(x).diff(x, 2) + exp(f(x)) + log(f(x)),
        'sol': [
            Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u - 2*exp(u)), (u, 1, f(x))), C2 + x),
            Eq(Integral(1/sqrt(C1 - 2*u*log(u) + 2*u - 2*exp(u)), (u, 1, f(x))), C2 - x),
        ],
        'simplify_flag': False,
    }
    return example_set
```
I think that's also fine. However we should be consistent across the different tests in the same file so unless we are planning to change *all* the examples to this format then I think the tests here should match what is done in the rest of the file.

@Mohitbalwani26 what do you think?

Mohitbalwani26 replied:
In my opinion, Only the example `separable_reduced_11` is too long rest all looks better to me as they are now. If we try to use your approach, Indention would be almost same as most of the examples look like this.
```python3
def _get_examples_ode_sol_separable_reduced():
    df = f(x).diff(x)
    return {
            'hint': "separable_reduced",
            'func': f(x),
            'examples':{
    'separable_reduced_01': {
        'eq': x* df  + f(x)* (1 / (x**2*f(x) - 1)),
        'sol': [Eq(log(x**2*f(x))/3 + log(x**2*f(x) - Rational(3, 2))/6, C1 + log(x))],
        'simplify_flag': False,
        'XFAIL': ['lie_group'], #It hangs.
    },
    }
    }
```

naveensaigit replied:
Yes, from what I observed, there was only 1 example that's too long. I'll look throughout the file change lines that are not too long into this format.

oscarbenjamin commented:
It is deliberate that we don't just call doit in checkodesol because it can be very slow for some equations. Internally doit should be called in some cases though.

What is the case that wasn't working in checkodesol?

Mohitbalwani26 commented:
@naveensaigit Also if you are fixing/making some change in the `checkodesol ` or `dsolve`, run `test_all_hints` function locally, because it might happen some xfail hints get fixed or new ones are to be added in the them.

naveensaigit commented:
> What is the case that wasn't working in checkodesol?

When `f(x)` is present inside an `Integral` like in this example - 
https://github.com/sympy/sympy/blob/4c20029861a0cdb8ab2356b6f57cc8f8d5b2b8b4/sympy/solvers/ode/tests/test_single.py#L385-L389
We need to do the doit, or else `checkodesol` doesn't do it and for some reason, the answer doesn't match the equation.

> @naveensaigit Also if you are fixing/making some change in the `checkodesol ` or `dsolve`, run `test_all_hints` function locally, because it might happen some xfail hints get fixed or new ones are to be added in the them.

Yes, currently I am running all tests using `sympy.test('sympy/solvers/ode')` and the doctests using `sympy.doctest('sympy/solvers/ode')`. This would include all the examples right?

Mohitbalwani26 commented:
> Yes, currently I am running all tests using `sympy.test('sympy/solvers/ode')` and the doctests using `sympy.doctest('sympy/solvers/ode')`. This would include all the examples right?

In test_single.py there is a function name `_test_all_hints()` change it to `test_all_hints` it will check all the examples to every hint then after runnig it locally change it back as it was.

oscarbenjamin commented:
> We need to do the doit, or else `checkodesol` doesn't do it and for some reason, the answer doesn't match the equation.

We could just mark that example as XFAIL for checkodesol. I'm sure there are improvements that can be made to checkodesol but I don't think that calling doit every time should be used.

It would be good to add an XFAIL test for checkodesol for this along with a test verifying that it works with `doit`.

naveensaigit commented:
I've removed doit and added tests, one with XFAIL and one with the doit. Any other changes required?

oscarbenjamin commented:
The tests for the autonomous solver are indented in a way that is inconsistent with the rest of the file. We discussed this before but it has been marked as resolved for some reason.

Otherwise this looks good.

@Mohitbalwani26 what do you think?

naveensaigit commented:
> The tests for the autonomous solver are indented in a way that is inconsistent with the rest of the file. We discussed this before but it has been marked as resolved for some reason.

I've actually changed the indentation of all examples with 2 or more solutions to the one you suggested. Where is it that you think its inconsistent? Do you mean the tests in other files like `test_ode.py`?

oscarbenjamin commented:
I'm talking about `_get_examples_ode_sol_2nd_nonlinear_autonomous_conserved` which has different indentation compared to every other similar function in `test_single.py`.

Mohitbalwani26 commented:
> The tests for the autonomous solver are indented in a way that is inconsistent with the rest of the file. We discussed this before but it has been marked as resolved for some reason. 
> @Mohitbalwani26 what do you think?

According to me also, `_get_examples_ode_sol_2nd_nonlinear_autonomous_conserved` looks different than rest. Also I see some solutions of examples are just re-written in new line, so if we are planning to write each solution equation on new line then it has to be done with rest of all the examples as well.

naveensaigit commented:
I've changed the indent to match the rest of the tests.

> Also I see some solutions of examples are just re-written in new line, so if we are planning to write each solution equation on new line then it has to be done with rest of all the examples as well.

I have changed this for all examples (with more than 1 solution) in `test_single.py`. Let me know if I missed anything.

oscarbenjamin commented:
> I've changed the indent to match the rest of the tests.

It still doesn't match.

naveensaigit commented:
> It still doesn't match.

I'm sorry, but I'm unable to understand where the difference is. Could you please suggest the changes?

oscarbenjamin commented:
These lines (like most examples in the file) are indented by 4, 8, and 12 spaces respectively:
https://github.com/sympy/sympy/blob/aefbc34872579ac2e701577ddd6f599dae1996ca/sympy/solvers/ode/tests/test_single.py#L1486-L1492

The examples that you have added e.g. `2nd_nonlinear_autonomous_conserved_01` are indented by 8, 12 and 16 spaces. They are indented one level more than all of the examples. Previously they were indented two levels more than the other examples and then you unindented them by one level but they still do not match the rest of the file.

naveensaigit commented:
Hopefully the indentation is correct now.

oscarbenjamin commented:
Yes, it looks good. Can you squash the commits can give a commit message like:
```
feat(dsolve): add a solver for some 2nd order autonomous ODEs
```

naveensaigit commented:
I've squashed all commits into one. Let me know if there's anything else.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,71 @@ def _get_general_solution(self, *, simplify: bool = True):
+        fy = self.wilds_match()[0]
+        u = Symbol('u')
+        syms = list(map(lambda sym:sym.name, fy.subs(fx, u).free_symbols))
+        if 'x' in syms:
```
This should not be comparing the string representation of symbols. Also what if the symbol name is not `x`?

naveensaigit replied:
So I should do?
``` python
x = self.ode_problem.sym
if x.name in syms:
.....
```

oscarbenjamin replied:
Yes, something like that

oscarbenjamin replied:
Wait no.

oscarbenjamin replied:
You should not be comparing the symbols by name. Symbols can have the same name but be different symbols e.g. if they have different assumptions or if they are Dummy.

You should compare the symbols directly so:
```
if x in syms:
```

naveensaigit replied:
I've made the changes. I'll rebase once you review again.

naveensaigit replied:
Also, I just realized that I don't need to type cast it into a `list`. I'll remove that whenever I push the changes next.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+        u = Symbol('u')
+        x = self.ode_problem.sym
+        syms = list(fy.subs(fx, u).free_symbols)
+        if x in syms:
```
What if the independent variable is called `u`? This should use a `Dummy`.

naveensaigit replied:
Ahh yes, why didn't I think of that!

oscarbenjamin commented:
> I'll rebase once you review again.

It isn't always necessary to rebase or squash the commits. However if the first approach tried in the PR ends up being completely rewritten (as was the case here) then we probably don't need the original commits and it is better to clean up the commit history.

The other reason for squashing is to improve the commit messages. Try to imagine what the changes look like in both the blame and the log. In the blame we see the file so we know at least which file is changed by each commit but the message should show the purpose of the change:
https://github.com/sympy/sympy/blame/9b99df776c03a892e26da86714a4ca10cb7efb84/sympy/solvers/ode/single.py
In the log we don't have that context so it's helpful if the message can give some context for which part of the codebase is being edited:
https://github.com/sympy/sympy/commits/master

Starting the commit message with `feat(dsolve)` means that we see that this is a feature being added to `dsolve` which is important contextual information. A message like "fix typo" is not very informative about why something was changed from a high-level.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+        fx = self.ode_problem.func
+        x = self.ode_problem.sym
+        u = Symbol('u')
+        g = g.subs(fx, u)
```
The `u` here should be a `Dummy` as well. What if `g` contains `u`?

naveensaigit commented:
Any other things to be fixed here?

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -784,5 +784,72 @@ def _get_general_solution(self, *, simplify: bool = True):
+        u = Dummy('u')
+        x = self.ode_problem.sym
+        syms = fy.subs(fx, u).free_symbols
+        if x in syms:
```
We should probably add this as a `is_autonomous` method on `ode_problem`.

naveensaigit replied:
Ah, yes, that is a nice idea. Would be useful later on.

naveensaigit replied:
Should I add `is_autonomous` to `SingleODEProblem` or `SinglePatternODESolver`? If it is the former, how would I be able to call the `wilds_match` as it wouldn't be defined?

oscarbenjamin replied:
I think it should be for `SingleODEProblem` as being autonomous is a property of the problem rather than a property of any solver.

Rather than using `wilds_match` we should think of a way to test for a generic ODE whether or not it is autonomous. I think that the technique that you use with `subs` here would probably work in general e.g.:
```python
In [7]: eq1 = f(x).diff(x, 2) + exp(f(x)) * f(x).diff(x)

In [8]: eq1.subs(f(x), y)
Out[8]: 
             2   
 y d        d    
ℯ ⋅──(y) + ───(y)
   dx        2   
           dx    

In [9]: x in eq1.subs(f(x), y).free_symbols
Out[9]: False

In [10]: eq2 = f(x).diff(x) + x*f(x)

In [11]: x in eq2.subs(f(x), y).free_symbols
Out[11]: True
```

naveensaigit replied:
I've done this. Should I add any tests for this? If so, where do you think they should go?

oscarbenjamin replied:
Are there already some tests for SingleODEProblem somewhere?

naveensaigit replied:
Yes, found them. I had overlooked them earlier.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -135,6 +135,14 @@ def iter_numbered_constants(self, start=1, prefix='C') -> Iterator[Symbol]:
             atom_set |= {Symbol(str(f.func)) for f in func_set}
         return numbered_symbols(start=start, prefix=prefix, exclude=atom_set)
 
+    def is_autonomous(self):
```
This could use `@cached_property`.

naveensaigit replied:
Anything else? I don't want the tests to run again just because of a small change :)

oscarbenjamin replied:
I left a couple more small comments but then it can be merged. Good work!

naveensaigit replied:
Thank you for your time to review this!

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -135,6 +135,14 @@ def iter_numbered_constants(self, start=1, prefix='C') -> Iterator[Symbol]:
+        syms = self.eq.subs(self.func, u).free_symbols
+        if x in syms:
+            return False
+        return True
```
```
return x not in syms
```

oscarbenjamin commented on sympy/solvers/ode/tests/test_single.py:
```diff
@@ -333,6 +334,12 @@ def test_SingleODESolver():
     assert problem.order == 4
 
+    problem = SingleODEProblem(f(x).diff(x, 3) + f(x).diff(x, 2) - f(x)**2, f(x), x)
+    assert problem.is_autonomous()
```
I would `assert ... is True`

Generally if you want to test `f` then `assert f()` is not a good test because almost every possible object in Python is truthy (e.g. `assert [1, 2]`).

oscarbenjamin commented:
This looks good.

I said I'd merge it so I will but I have also thought of an improvement which is that it might make sense to do the `is_autonomous` checking before calling `match` rather than in the `_verify` method. If we had many solvers for autonomous equations that would definitely be better (the solvers could have an `autonomous` property to be checked).

naveensaigit commented:
Are there any classes of autonomous equations that cover a great range of ODEs?

oscarbenjamin commented:
> Are there any classes of autonomous equations that cover a great range of ODEs?

Not that I know of but autonomous ODEs are very common in applications.

oscarbenjamin commented:
Actually there is https://github.com/sympy/sympy/issues/4414#issuecomment-753918012

naveensaigit commented:
There is also a special case of 2nd order autonomous ode which can be solved exactly. [Wikipedia](https://en.wikipedia.org/wiki/Autonomous_system_(mathematics)#Special_case:_x''_=_x'n_f(x)). Currently, sympy cannot solve this - 
``` python
>>> from sympy import *
>>> f=Function('f')
>>> x=symbols('x')
>>> eq = f(x).diff(x,2)-f(x).diff(x)*f(x)
>>> dsolve(eq)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Naveen S Thota\sympy\sympy\solvers\ode\ode.py", line 629, in dsolve
    x0=x0, n=n, **kwargs)
  File "C:\Users\Naveen S Thota\sympy\sympy\solvers\deutils.py", line 241, in _desolve
    raise NotImplementedError(dummy + "solve" + ": Cannot solve " + str(eq))
NotImplementedError: solve: Cannot solve -f(x)*Derivative(f(x), x) + Derivative(f(x), (x, 2))
```
This might be a worthy addition.

oscarbenjamin commented:
> This might be a worthy addition.

Yes, I think so. In fact isn't that a generalisation of the class of ODE handled here?

naveensaigit commented:
Yes, I should've noticed this earlier. That way, a lot of time would be saved. But I don't think adding this will take much time. I will only have to change a few things like the `Integral` inside the solver, add some documentation and some tests. 
Maybe we should rename the hint to something without "conserved". What do you suggest?

oscarbenjamin commented:
I'm not sure what a good name is. Generally what this is doing is finding a "first integral" but that name would be confusing compared to the other hints as it would imply that it is 1st order and that you want answers in terms of Integral.
